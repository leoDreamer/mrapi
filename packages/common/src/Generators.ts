import type { mrapi } from './types'

import { join } from 'path'
import { existsSync, mkdirSync, writeFileSync } from 'fs'

import { formation } from './shell'

// const tscOptions = [
//   '-t es2018',
//   '--lib esnext',
//   '--module commonjs',
//   '--moduleResolution node',
//   '--allowSyntheticDefaultImports',
//   '--esModuleInterop',
//   '--importHelpers',
//   '--resolveJsonModule',
//   '--sourceMap false ',
//   '--declaration',
//   '--skipLibCheck',
// ].join(' ')

export class Generators {
  protected options: mrapi.generate.Options = {
    schema: '',
    output: '',
    excludeFields: [],
    excludeModels: [],
    excludeFieldsByModel: {},
    excludeQueriesAndMutations: [],
    excludeQueriesAndMutationsByModel: {},
  }
  protected isJS?: boolean = false
  protected queries: mrapi.generate.Query[] = [
    'findOne',
    'findMany',
    'findCount',
    'aggregate',
  ]
  protected mutations: mrapi.generate.Mutation[] = [
    'createOne',
    'updateOne',
    'upsertOne',
    'deleteOne',
    'updateMany',
    'deleteMany',
  ]

  constructor(customOptions?: Partial<mrapi.generate.Options>) {
    this.options = { ...this.options, ...customOptions }
    this.isJS = this.options.javascript
  }

  protected async dmmf() {
    const { dmmf } = await import(this.options.schema)
    return dmmf
  }

  protected async datamodel() {
    const { datamodel }: { datamodel: any } = await this.dmmf()
    return datamodel
  }

  protected async models() {
    const { schema }: { schema: any } = await this.dmmf()
    // Plan to replace "outputTypes" to "datamodel.models"
    return schema.outputTypes.filter(
      (model: any) =>
        !['Query', 'Mutation'].includes(model.name) &&
        !model.name.includes('Aggregate') &&
        model.name !== 'BatchPayload' &&
        (!this.options.models || this.options.models.includes(model.name)),
    )
  }

  protected withExtension(filename: string) {
    return filename + (this.isJS ? '.js' : '.ts')
  }

  protected excludeFields(model: string) {
    return this.options.excludeFields.concat(
      this.options.excludeFieldsByModel[model],
    )
  }

  protected disableQueries(model: string) {
    return (
      !this.options.disableQueries &&
      !this.options.excludeModels.find(
        (item) => item.name === model && item.queries,
      )
    )
  }

  protected disableMutations(model: string) {
    return (
      !this.options.disableMutations &&
      !this.options.excludeModels.find(
        (item) => item.name === model && item.mutations,
      )
    )
  }

  protected smallModel(name: string) {
    return name.charAt(0).toLowerCase() + name.slice(1)
  }

  protected excludedOperations(model: string) {
    return this.options.excludeQueriesAndMutations.concat(
      this.options.excludeQueriesAndMutationsByModel[model] ?? [],
    )
  }

  protected mkdir(path: string) {
    !existsSync(path) && mkdirSync(path, { recursive: true })
  }

  protected output(...paths: string[]): string {
    return join(this.options.output, ...paths)
  }

  protected getIndexContent(files: string[]) {
    const lines: string[] = []
    if (this.isJS) lines.push('module.exports = {')
    files.forEach((file) => {
      if (this.isJS) {
        lines.push(`  ...require('./${file}'),`)
      } else {
        lines.push(`export * from './${file}'`)
      }
    })
    if (this.isJS) lines.push('}')
    return lines.join('\n')
  }

  protected getImport(content: string, path: string) {
    return this.isJS
      ? `const ${content} = require('${path}')`
      : `import ${content} from '${path}'`
  }

  protected createFileIfNotfound(
    path: string,
    fileName: string,
    content: string,
  ) {
    !existsSync(path) && this.mkdir(path)
    !existsSync(join(path, fileName)) &&
      writeFileSync(join(path, fileName), content)
  }

  protected get parser() {
    return this.isJS ? 'babel' : 'babel-ts'
  }

  // protected formation = formation
  protected formation(text: string) {
    return formation(text, this.parser)
  }

  // async toJS() {
  //   const { output } = this.options

  //   // Glob is blocked in windows. That is `npx tsc **/*.ts` cannot execute in windows. It maybe a problem of npx.
  //   // The script underline cannot execute successfully in windows.
  //   // const exitPalCode = await spawnShell(
  //   //   `npx tsc ${tscOptions} ${output}/*.ts ${output}/**/*.ts ${output}/**/**/*.ts`,
  //   // )

  //   const files: string[] = [
  //     ...glob.sync(`${output}/*.ts`),
  //     ...glob.sync(`${output}/**/*.ts`),
  //     ...glob.sync(`${output}/**/**/*.ts`),
  //   ]
  //   const exitPalCode = await spawnShell(
  //     `npx tsc ${tscOptions} ${files.join(' ')}`,
  //   )
  //   if (exitPalCode !== 0) {
  //     throw new Error('Generate nexus types exception.')
  //   }
  // }
}
