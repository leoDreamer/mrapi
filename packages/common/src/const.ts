export const FILE_HEADER = `/**
* This file was generated by mrapi
* Do not make changes to this file directly
*/
`
