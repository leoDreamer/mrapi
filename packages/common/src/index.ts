import merge from 'deepmerge'

export * from './config'
export * from './utils'
export { merge }
