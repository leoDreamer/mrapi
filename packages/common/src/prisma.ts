export default function getPrisma(prismaClient: any) {
  if (!prismaClient) {
    throw new Error(`[/${prismaClient}] - PrismaClient was not found.`)
  }
  return typeof prismaClient === 'string' ? require(prismaClient) : prismaClient
}

export function getPrismaClient(prismaClient: any) {
  return getPrisma(prismaClient).PrismaClient
}

export function getPrismaDmmf(prismaClient: any) {
  return getPrisma(prismaClient).dmmf
}
