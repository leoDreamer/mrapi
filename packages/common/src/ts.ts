import { dirname } from 'path'
import { tryRequire } from './utils'
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs'

export function compileTSFile(src: string, dest: string) {
  if (!existsSync(src)) {
    return ''
  }

  const {
    transpileModule,
    ModuleKind,
  }: typeof import('typescript') = tryRequire(
    'typescript',
    'Please install it manually.',
  )

  const source = readFileSync(src, { encoding: 'utf-8' })
  const result = transpileModule(source, {
    compilerOptions: { module: ModuleKind.CommonJS },
  })

  if (!result || !result.outputText) {
    return ''
  }

  const destDir = dirname(dest)
  !existsSync(destDir) && mkdirSync(destDir, { recursive: true })
  writeFileSync(dest, result.outputText)

  return dest
}
