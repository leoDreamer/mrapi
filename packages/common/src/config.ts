import type mrapi from '@mrapi/types'

import assert from 'assert'
import merge from 'deepmerge'
import { existsSync } from 'fs'
import { compileTSFile } from './ts'
import { join, dirname, relative } from 'path'
import { ensureAbsolutePath, tryRequire, getWorkspaceDirs } from './utils'

process.env.NODE_ENV = process.env.NODE_ENV || 'development'
const isProd = process.env.NODE_ENV === 'production'
const workDirs = getWorkspaceDirs()
const src = isProd ? workDirs.dst : workDirs.src

interface PathsObject {
  output: string
  custom: string
}

const defaultPrismaOptions = {
  output: 'node_modules/.prisma/client',
  schema: 'prisma/schema.prisma',
}

const defaultApiOutput = 'node_modules/.mrapi/'
const defaultPrismaOutput = 'node_modules/.prisma/'

const defaultServiceConfig: Partial<mrapi.ServiceOptions> = {
  defaultTenant: '',
  tenantIdentity: 'mrapi-tenant-id',
  studio: !isProd,
}

// management is a special service which has no 'apiOutput'
const defaultManagementConfig: Partial<mrapi.ServiceOptions> = {
  ...defaultServiceConfig,
  managementTenantModelName: 'tenant',
}

export const defaults = {
  config: {
    cwd: process.cwd(),
    service: defaultServiceConfig,
    autoGenerate: true,
    isMultiService: false,
  },
  clientPath: join(process.cwd(), 'node_modules/@prisma/client'),
  configFileName: 'mrapi.config',
  tenantName: 'default',
  serviceName: 'default',
  port: 1358,
  graphql: {
    playground: !isProd,
    generator: 'nexus' as mrapi.GraphqlGenerator,
  },
  openapi: false,
}

export function resolveConfig(
  input?: mrapi.ConfigInput | mrapi.Config,
  cwd = process.cwd(),
  configFileName = defaults.configFileName,
): mrapi.Config {
  if (input?.parsed) {
    return input as mrapi.Config
  }

  const maybeConfigPath =
    process.env.MRAPI_CONFIG_PATH ||
    join(cwd, configFileName || defaults.configFileName)
  const configPath = resolveConfigFilePath(maybeConfigPath)

  if (!configPath) {
    throw new Error(
      `Can not resolve config file from path: ${relative(
        cwd,
        maybeConfigPath,
      )}`,
    )
  }

  const tmp = tryRequire(configPath)
  const config: mrapi.ConfigInput = merge(
    {
      ...defaults.config,
      ...(tmp || {}),
    },
    (input || {}) as mrapi.Config,
  )
  const ServiceCwd = dirname(maybeConfigPath)
  const isMultiService = Array.isArray(config.service)
  const services: Array<mrapi.ServiceOptionsInput> = config.service
    ? Array.isArray(config.service)
      ? config.service
      : [config.service]
    : []

  if (isMultiService) {
    const hasName = services.every(
      (val: mrapi.ServiceOptionsInput) => !!val.name,
    )
    assert(
      hasName,
      `[Config Error] Multiple services should have 'name' fields on each.`,
    )
  }

  const service = services.map((s: mrapi.ServiceOptionsInput) =>
    normalizeServiceConfig(s, { isMultiService, cwd: ServiceCwd }),
  )

  return {
    ...config,
    service,
    cwd: ServiceCwd,
    isMultiService,
    autoGenerate:
      config.autoGenerate !== undefined
        ? config.autoGenerate
        : defaults.config.autoGenerate,
    parsed: true,
  }
}

function resolveConfigFilePath(configPath: string): string {
  const tsFile = configPath + '.ts'
  if (existsSync(tsFile)) {
    const destFile = compileTSFile(
      tsFile,
      ensureAbsolutePath(join(defaultApiOutput, 'config.js')),
    )

    return destFile
  }

  const jsFile = configPath + '.js'
  if (existsSync(jsFile)) {
    return jsFile
  }

  return ''
}

function normalizeServiceConfig(
  service: mrapi.ServiceOptionsInput,
  { isMultiService, cwd }: { isMultiService: boolean; cwd: string },
): mrapi.ServiceOptions {
  service.name = service.name || defaults.serviceName

  if (service.tenants) {
    assert(
      Array.isArray(service.tenants),
      `[Config Error] 'service.tenants' should be an array `,
    )
  }

  const isMultiTenant = Array.isArray(service.tenants)

  // datasource paths
  const usingDatasource = service.datasource || service.schema
  if (usingDatasource) {
    const hasDatabase = isMultiTenant
      ? service.tenants?.every((t) => !!t.database)
      : !!service.database
    assert(
      hasDatabase,
      `[Config Error] Service '${
        service.name
      }' using prisma, but no 'database' field configured.${
        isMultiTenant
          ? `Each tenant should configure 'database' field when using multi-tenant`
          : ''
      }`,
    )

    service.datasource = {
      ...service.datasource,
      provider:
        service.datasource?.provider || ('prisma' as mrapi.DatasourceProvider),
      schema: ensureAbsolutePath(
        service.datasource?.schema ||
          service.schema ||
          defaultPrismaOptions.schema,
        cwd,
      ),
      output: ensureAbsolutePath(
        service.datasource?.output ||
          join(
            defaultPrismaOutput,
            `${isMultiService ? service.name + '-' : ''}client`,
          ),
        cwd,
      ),
    }
  } else {
    delete service.datasource
  }

  // custom source path
  service.customDir = ensureAbsolutePath(
    service.customDir || `${src}${isMultiService ? `/${service.name}` : ''}`,
  )
  const contextFile = join(service.customDir, 'context')

  if (isMultiTenant) {
    service.tenants = service.tenants?.map((t) => ({
      ...t,
      name: t.name || defaults.tenantName,
    }))
  } else {
    delete service.tenants
  }

  const defaultConfig = JSON.parse(
    JSON.stringify(
      service.management ? defaultManagementConfig : defaultServiceConfig,
    ),
  )

  return merge(defaultConfig, {
    ...(service || {}),
    graphql: normalizeGraphqlConfig(service, {
      isMultiService,
      cwd,
    }),
    openapi: normalizeOpenapiConfig(service, {
      isMultiService,
      cwd,
    }),
    isMultiTenant,
    contextFile,
  })
}

function normalizeGraphqlConfig(
  service: mrapi.ServiceOptionsInput,
  { isMultiService, cwd }: { isMultiService: boolean; cwd: string },
): mrapi.GraphqlOptions | undefined {
  if (service.graphql === false) {
    return undefined
  }

  const tmp = (service.graphql || {}) as PathsObject
  const options = merge(defaults.graphql, tmp)

  return {
    ...options,
    output: ensureAbsolutePath(
      tmp?.output ||
        join(defaultApiOutput, isMultiService ? service.name! : '', 'graphql'),
      cwd,
    ),
    custom: ensureAbsolutePath(join(service.customDir!, 'graphql'), cwd),
    playground: !!options.playground,
  }
}

function normalizeOpenapiConfig(
  service: mrapi.ServiceOptionsInput,
  { isMultiService, cwd }: { isMultiService: boolean; cwd: string },
): mrapi.OpenapiOptions | undefined {
  if (!service.openapi) {
    return undefined
  }

  const tmp = (service.openapi || {}) as PathsObject
  const options = merge(defaults.openapi, tmp)

  return {
    ...options,
    output: ensureAbsolutePath(
      tmp?.output ||
        join(defaultApiOutput, isMultiService ? service.name! : '', 'openapi'),
      cwd,
    ),
    custom: ensureAbsolutePath(join(service.customDir!, 'openapi'), cwd),
  }
}
