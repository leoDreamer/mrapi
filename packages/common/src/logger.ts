import type { LoggerOptions } from 'pino'
import pino, { Logger } from 'pino'

import { defaultLoggerOptions } from './config'

function createLogger(options: LoggerOptions): Logger {
  return pino(options)
}

function getLogger(logger: null | Logger, opts?: LoggerOptions): Logger {
  return (
    logger ||
    createLogger({
      ...defaultLoggerOptions,
      ...opts,
    })
  )
}

const logger = createLogger(defaultLoggerOptions)

export { Logger, LoggerOptions, logger, getLogger }
