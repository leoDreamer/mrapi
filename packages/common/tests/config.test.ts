import { join } from 'path'
import { resolveConfig } from '../src/index'

const fixturesRoot = join(__dirname, './__fixtures__/config')

describe('Config', () => {
  beforeEach(() => {
    jest.resetModules()
    process.env.NODE_ENV ='production'
  })

  afterEach(() => {})
  test('types', () => {
    expect(typeof resolveConfig).toBe('function')
  })

  test('no-config', () => {
    const cwd = join(fixturesRoot, 'no-config')
    const config = resolveConfig({}, cwd)
    expect(typeof config).toBe('object')

    expect(config).toMatchSnapshot()
  })

  test('basic', () => {
    const cwd = join(fixturesRoot, 'configs')
    const config = resolveConfig({}, cwd, 'mrapi.config.basic')

    expect(typeof config).toBe('object')
    expect(config).toMatchSnapshot()
  })

  test('multi-tenant', () => {
    const cwd = join(fixturesRoot, 'configs')
    const config = resolveConfig({}, cwd, 'mrapi.config.multi-tenant')

    expect(typeof config).toBe('object')
    expect(config).toMatchSnapshot()
  })

  test('multi-service', () => {
    const cwd = join(fixturesRoot, 'configs')
    const config = resolveConfig({}, cwd, 'mrapi.config.multi-service')

    expect(typeof config).toBe('object')
    expect(config).toMatchSnapshot()
  })

  test('disable-api', () => {
    const cwd = join(fixturesRoot, 'configs')
    const config = resolveConfig({}, cwd, 'mrapi.config.disable-api')

    expect(typeof config).toBe('object')
    expect(config).toMatchSnapshot()
  })

  test('no-database', () => {
    const cwd = join(fixturesRoot, 'configs')
    try {
      const config = resolveConfig({}, cwd, 'mrapi.config.no-database')
      console.log(JSON.stringify(config, null, 2))
    } catch (err) {
      expect(err.message).toContain(
        `using prisma, but no 'database' field configured`,
      )
    }
  })

  test('multi-service no-name', () => {
    const cwd = join(fixturesRoot, 'configs')
    try {
      resolveConfig({}, cwd, 'mrapi.config.no-name')
    } catch (err) {
      expect(err.message).toContain(
        `Multiple services should have 'name' fields on each`,
      )
    }
  })
  test('resolveConfig() input.parsed exsits', () => {
    const cwd = join(fixturesRoot, 'configs')
    // input.parsed exsits
    const out = resolveConfig({ parsed: true }, cwd, 'mrapi.config.basic')
    expect(out).toEqual({ parsed: true })
    const out1 = resolveConfig({ parsed: false }, cwd, 'mrapi.config.basic')
    expect(out1.parsed).toBeTruthy()
    const out2 = resolveConfig(undefined, cwd, 'mrapi.config.basic')
    expect(out2.parsed).toBeTruthy()
    //  existsSync(tsFile)
    try {
      const out1 = resolveConfig({}, cwd, 'this.is.no-match')
      expect(out1).toReturn()
      expect(out1.service.length).toBe(1)
      expect(out1.isMultiService).toBeFalsy()
      expect(typeof out1.cwd).toBe('string')
      expect(out1.parsed).toBeTruthy()
    } catch (error) {}
  })
  test('process.env.NODE_ENV ', () => {
    const cwd = join(fixturesRoot, 'configs')
    // production
    process.env.NODE_ENV ='production'
    const out=resolveConfig({}, cwd, 'mrapi.config.basic')
    expect(process.env.NODE_ENV).toBe('production')
    expect(typeof out).toBe('object')
    // null
    delete process.env.NODE_ENV
    const out2=resolveConfig({}, cwd, 'mrapi.config.basic')
    expect(process.env.NODE_ENV).toBeUndefined()
    expect(typeof out2).toBe('object')
    // test
    process.env.NODE_ENV='test'
    const out3=resolveConfig({}, cwd, 'mrapi.config.basic')
    expect(process.env.NODE_ENV).toBe('test')
    expect(typeof out3).toBe('object')
  })
  test('normalizeOpenapiConfig() service.tenants ', () => {
    const cwd = join(fixturesRoot, 'configs')
    // service.tenants is not an array
    const input:any={service:{tenants:'jfasnjs'}}
    try {
      resolveConfig(input, cwd, 'mrapi.config.basic')
    } catch (err) {
      expect(err.message).toContain(
        "[Config Error] 'service.tenants' should be an array ",
      )
    }
    // service.tenants is an array and it don't have db
    try {
      resolveConfig({service:{tenants:[{name:"string"}]}}, cwd, 'mrapi.config.basic')
    } catch (err) {
      expect(err.message).toContain(
        "[Config Error] Service 'default' using prisma, but no 'database' field configured. ",
      )
    }

    try {
      resolveConfig({service:{tenants:[{name:"string"}],multiTenant:{mode:'seprate-db'}}}, cwd, 'mrapi.config.basic')
    } catch (err) {
      expect(err.message).toContain(
        "Each tenant should configure \'database\' field when using multi-tenant \'seprate-db\' mode",
      )
    }
    
   
    
  })
  test('service.datasource.provider ', () => {
    const cwd = join(fixturesRoot, 'configs')
    const out= resolveConfig({}, cwd, 'mrapi.config.multi-tenant-db-provider')
    expect(typeof out).toBe('object')
    expect(typeof out.service).toBe('object')
    const out1= resolveConfig({}, cwd, 'mrapi.config.multi-tenant-true')
    expect(typeof out1).toBe('object')
    expect(typeof out1.service).toBe('object')
    // openapi is not exists
    const out2=resolveConfig({},cwd,'mrapi.config.openapi-is-false')
    expect(typeof out2).toBe('object')
  })
  test('resolveConfig no params ', () => {
      let nullValue:any
      const cwd = join(fixturesRoot, 'configs')
      const nullConfigFileName:any=null
      const notExistFileName='mrapi.config.not-exsits-file'
      const notExistConfigService='mrapi.config.no-config-service'
      const out= resolveConfig(nullValue)
      expect(typeof out).toBe('object')
      // test configFileName is null
      const out1=resolveConfig(nullValue,nullValue,nullConfigFileName)
      expect(typeof out1).toBe('object')
      // configFileName is not exist
      const out2=resolveConfig(nullValue,nullValue,notExistFileName)
      expect(typeof out2).toBe('object')
      const out3=resolveConfig(nullValue,cwd,notExistFileName)
      expect(typeof out3).toBe('object')
      // config-service is not exist
      const out4=resolveConfig(nullValue,cwd,notExistConfigService)
      expect(typeof out4).toBe('object')
  });
  test('no-schema', () => {
    const cwd = join(fixturesRoot, 'configs')
    const out= resolveConfig({},cwd,'mrapi.config.no-schema')
    expect(typeof out).toBe('object')
  });
    test('multi-tenant-no-name', () => {
      const cwd = join(fixturesRoot, 'configs')
    const out= resolveConfig({},cwd,'mrapi.config.multi-tenant-no-name')
    expect(typeof out).toBe('object')
    });
    test('openapi-multi-false', () => {
      const cwd = join(fixturesRoot, 'configs')
      const out= resolveConfig({},cwd,'mrapi.config.openapi-multi-false')
      expect(typeof out).toBe('object')
    });
})
