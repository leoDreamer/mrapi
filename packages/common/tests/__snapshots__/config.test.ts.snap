// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config basic 1`] = `
Object {
  "__cwd": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs",
  "__isMultiService": false,
  "__parsed": true,
  "autoGenerate": true,
  "service": Array [
    Object {
      "__isMultiTenant": false,
      "database": "file:./dev.db",
      "datasource": Object {
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.prisma/client",
        "provider": "prisma",
        "schema": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/prisma/schema.prisma",
      },
      "defaultTenant": "",
      "graphql": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/graphql",
        "generator": "nexus",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/graphql",
        "playground": true,
      },
      "name": "default",
      "openapi": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/openapi",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/openapi",
      },
      "schema": "prisma/schema.prisma",
      "studio": true,
      "tenantIdentity": "mrapi-tenant-id",
    },
  ],
}
`;

exports[`Config disable-api 1`] = `
Object {
  "__cwd": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs",
  "__isMultiService": true,
  "__parsed": true,
  "autoGenerate": true,
  "service": Array [
    Object {
      "__isMultiTenant": true,
      "datasource": Object {
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.prisma/user-client",
        "provider": "prisma",
        "schema": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/prisma/user/schema.prisma",
      },
      "defaultTenant": "one",
      "graphql": false,
      "name": "user",
      "openapi": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/openapi/user",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/user/openapi",
      },
      "schema": "prisma/user/schema.prisma",
      "studio": true,
      "tenantIdentity": "mrapi-tenant-id",
      "tenants": Array [
        Object {
          "database": "file:./user-one.db",
          "name": "one",
        },
        Object {
          "database": "file:./user-two.db",
          "name": "two",
        },
      ],
    },
    Object {
      "__isMultiTenant": true,
      "database": "file:./post.db",
      "datasource": Object {
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.prisma/post-client",
        "provider": "prisma",
        "schema": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/prisma/post/schema.prisma",
      },
      "defaultTenant": "one",
      "graphql": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/graphql/post",
        "generator": "nexus",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/post/graphql",
        "playground": true,
      },
      "name": "post",
      "openapi": false,
      "schema": "prisma/post/schema.prisma",
      "studio": true,
      "tenantIdentity": "mrapi-tenant-id",
      "tenants": Array [
        Object {
          "database": "file:./post-one.db",
          "name": "one",
        },
        Object {
          "database": "file:./post-two.db",
          "name": "two",
        },
      ],
    },
    Object {
      "__isMultiTenant": false,
      "database": "file:./dev.db",
      "datasource": Object {
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.prisma/management-client",
        "provider": "prisma",
        "schema": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/prisma/management/schema.prisma",
      },
      "defaultTenant": "",
      "graphql": false,
      "management": true,
      "managementTenantModelName": "tenant",
      "name": "management",
      "openapi": false,
      "schema": "prisma/management/schema.prisma",
      "studio": true,
      "tenantIdentity": "mrapi-tenant-id",
    },
  ],
}
`;

exports[`Config multi-service 1`] = `
Object {
  "__cwd": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs",
  "__isMultiService": true,
  "__parsed": true,
  "autoGenerate": true,
  "service": Array [
    Object {
      "__isMultiTenant": true,
      "datasource": Object {
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.prisma/user-client",
        "provider": "prisma",
        "schema": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/prisma/user/schema.prisma",
      },
      "defaultTenant": "one",
      "graphql": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/graphql/user",
        "generator": "nexus",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/user/graphql",
        "playground": true,
      },
      "name": "user",
      "openapi": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/openapi/user",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/user/openapi",
      },
      "schema": "prisma/user/schema.prisma",
      "studio": true,
      "tenantIdentity": "mrapi-tenant-id",
      "tenants": Array [
        Object {
          "database": "file:./user-one.db",
          "name": "one",
        },
        Object {
          "database": "file:./user-two.db",
          "name": "two",
        },
      ],
    },
    Object {
      "__isMultiTenant": true,
      "database": "file:./post.db",
      "datasource": Object {
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.prisma/post-client",
        "provider": "prisma",
        "schema": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/prisma/post/schema.prisma",
      },
      "defaultTenant": "one",
      "graphql": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/graphql/post",
        "generator": "nexus",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/post/graphql",
        "playground": true,
      },
      "name": "post",
      "openapi": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/openapi/post",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/post/openapi",
      },
      "schema": "prisma/post/schema.prisma",
      "studio": true,
      "tenantIdentity": "mrapi-tenant-id",
      "tenants": Array [
        Object {
          "database": "file:./post-one.db",
          "name": "one",
        },
        Object {
          "database": "file:./post-two.db",
          "name": "two",
        },
      ],
    },
    Object {
      "__isMultiTenant": false,
      "database": "file:./dev.db",
      "datasource": Object {
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.prisma/management-client",
        "provider": "prisma",
        "schema": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/prisma/management/schema.prisma",
      },
      "defaultTenant": "",
      "graphql": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/graphql/management",
        "generator": "nexus",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/management/graphql",
        "playground": true,
      },
      "management": true,
      "managementTenantModelName": "tenant",
      "name": "management",
      "openapi": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/openapi/management",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/management/openapi",
      },
      "schema": "prisma/management/schema.prisma",
      "studio": true,
      "tenantIdentity": "mrapi-tenant-id",
    },
  ],
}
`;

exports[`Config multi-tenant 1`] = `
Object {
  "__cwd": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs",
  "__isMultiService": false,
  "__parsed": true,
  "autoGenerate": true,
  "service": Array [
    Object {
      "__isMultiTenant": true,
      "database": "file:./dev.db",
      "datasource": Object {
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.prisma/client",
        "provider": "prisma",
        "schema": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/prisma/schema.prisma",
      },
      "defaultTenant": "",
      "graphql": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/graphql",
        "generator": "nexus",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/graphql",
        "playground": true,
      },
      "name": "default",
      "openapi": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/src/openapi",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/configs/node_modules/.mrapi/openapi",
      },
      "schema": "prisma/schema.prisma",
      "studio": true,
      "tenantIdentity": "mrapi-tenant-id",
      "tenants": Array [
        Object {
          "database": "file:./user-one.db",
          "name": "one",
        },
        Object {
          "database": "file:./user-two.db",
          "name": "two",
        },
      ],
    },
  ],
}
`;

exports[`Config no-config 1`] = `
Object {
  "__cwd": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/no-config",
  "__isMultiService": false,
  "__parsed": true,
  "autoGenerate": true,
  "service": Array [
    Object {
      "__isMultiTenant": false,
      "defaultTenant": "",
      "graphql": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/no-config/src/graphql",
        "generator": "nexus",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/no-config/node_modules/.mrapi/graphql",
        "playground": true,
      },
      "name": "default",
      "openapi": Object {
        "custom": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/no-config/src/openapi",
        "output": "<PROJECT_ROOT>/packages/common/tests/__fixtures__/config/no-config/node_modules/.mrapi/openapi",
      },
      "studio": true,
      "tenantIdentity": "mrapi-tenant-id",
    },
  ],
}
`;
